### PR TITLE
[Model] Support Dots OCR

### LINKED ISSUE #24581: [New Model]: dots_ocr by rednote
### The model to consider.

https://huggingface.co/rednote-hilab/dots.ocr

Modeling code exists: https://huggingface.co/rednote-hilab/dots.ocr/blob/main/modeling_dots_ocr_vllm.py

### The closest model vllm already supports.

Qwen2.5 VL

### What's your difficulty of supporting the model you want?

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by ywang96**
## Purpose
This PR adds support for [rednote-hilab/dots.ocr](https://huggingface.co/rednote-hilab/dots.ocr). This model is currently supported via OOT registration but we might as well bring it into vLLM so that users don't need to set it up with additonal steps.

Most of the codes taken are from the model repo, but this PR also cleans up a few logics that are no longer needed since this model does not support video modality.

FIXES #24581 

Co-authored-by @yinz-aizip 

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by casper-hansen** [comment]
Thanks for taking a stab at upstreaming this @ywang96. We need more and better OCR models, and this would be a great step forward.

**Comment by ywang96** [comment]
> Thanks for taking a stab at upstreaming this @ywang96. We need more and better OCR models, and this would be a great step forward.

@casper-hansen No problem. BTW there were still some performance issues I still need to debug, but correctness-wise, this branch should be actually ready to go with `vllm serve rednote-hilab/dots.ocr --trust-remote-code`.

**Comment by ywang96** [comment]
@yinz-aizip I made some changes to use vllm internal layers - could you help verify the correctness of this implementation (similarly to what you did in your PR)? Thanks!

**Comment by yinz-aizip** [comment]
## Summary

This PR compares performance and evaluation metrics between two commits:
	•	Commit 96cc9bc40f464d16ddec653945e28476decf1ed1
	•	Commit c830194f4a5c67eb0725799e65ced7b06a5824bd

The model was served with the following configuration:

``` bash
hf_model_path='/path/to/dots.ocr'
export CUDA_VISIBLE_DEVICES=4,5,6,7
vllm serve $hf_model_path \
    --host 127.0.0.1 \
    --port 8126 \
    --data-parallel-size 4 \
    --tensor-parallel-size 1 \
    --gpu-memory-utilization 0.8 \
    --chat-template-content-format string \
    --served-model-name model \
    --trust-remote-code
```

## Efficiency

Throughput was estimated by running 1,000 concurrent requests on a single image.
	•	Commit 96cc9bc40f4... → 3980.84
	•	Commit c830194f4a5... → 3743.34

The newer commit is slightly slower, though the difference is relatively minor.

## Effectiveness

Evaluated on OmniDocBenchmark (lower is better):
	•	Commit 96cc9bc40f4...
	•	overall_EN: 0.12578
	•	overall_CH: 0.16556
	•	Commit c830194f4a5...
	•	overall_EN: 0.12412
	•	overall_CH: 0.16285

The newer commit shows a small improvement in accuracy across both English and Chinese benchmarks.

## Conclusion
	•	Efficiency: Slightly reduced in the newer commit.
	•	Effectiveness: Marginally improved results on OmniDocBenchmark.

Overall, the trade-off seems acceptable, with minor throughput loss balanced by better benchmark performance.

**Comment by casper-hansen** [comment]
@ywang96 I tested this PR:

- [x] Correctness issues fixed (no more endless generations)
- [x] Performance issues also fixed, now scales with number of concurrent images.

1x H100 concurrency benchmark:
- 1.81 images/second for 16 images.
- 2.36 images/second for 30 images.

**Comment by casper-hansen** [comment]
One potential performance issue: When I pass in 30 images, I see the below message 30 times before images are actually processed. This seems to add a big latency for this model.

```
Fetching 1 files: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 7869.24it/s]
```

**Comment by ywang96** [comment]
> One potential performance issue: When I pass in 30 images, I see the below message 30 times before images are actually processed. This seems to add a big latency for this model.
> 
> ```
> Fetching 1 files: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 7869.24it/s]
> ```

Hmm - FYI @DarkLight1337, I'm wondering if this has something to do with a out-of-tree processor (i.e one with `trust_remote_code=True`) inheriting from a HF one?

File: vllm/model_executor/models/registry.py [hunk 1]
```diff
@@ -213,6 +213,7 @@
     "ChameleonForConditionalGeneration": ("chameleon", "ChameleonForConditionalGeneration"),  # noqa: E501
     "Cohere2VisionForConditionalGeneration": ("cohere2_vision", "Cohere2VisionForConditionalGeneration"),  # noqa: E501
     "DeepseekVLV2ForCausalLM": ("deepseek_vl2", "DeepseekVLV2ForCausalLM"),
+    "DotsOCRForCausalLM": ("dots_ocr", "DotsOCRForCausalLM"),
```
Comment (jeejeelee) [inline vllm/model_executor/models/registry.py:222]: We aslo need to add this model in https://github.com/vllm-project/vllm/blob/main/tests/models/registry.py

File: vllm/model_executor/models/dots_ocr.py [hunk 2]
```diff
@@ -0,0 +1,829 @@
...
+                                        hidden_features,
+                                        bias=bias,
+                                        quant_config=quant_config,
+                                        prefix=f"{prefix}.fc3",
+                                        disable_tp=True)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x1, _ = self.fc1(x)
+        x3, _ = self.fc3(x)
+        x = F.silu(x1) * x3
```
Comment (jeejeelee) [inline vllm/model_executor/models/dots_ocr.py]: Can we use `MergedColumnParallelLinear` here ?

File: vllm/model_executor/models/dots_ocr.py [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+            raise RuntimeError("Unsupported attention backend")
+
+        # [B,S,H,D] -> [S,B,H*D] -> [S, C]
+        context_layer = context_layer.permute(1, 0, 2, 3).contiguous()
+        context_layer = context_layer.view(context_layer.shape[0], bs, -1)
+        out, _ = self.proj(context_layer)
+        return out.squeeze(1)
+
+
+class DotsSwiGLUFFN(nn.Module):
```
Comment (jeejeelee) [inline vllm/model_executor/models/dots_ocr.py:365]: It looks like we can use `Idefics2VisionMLP` directly

File: vllm/model_executor/models/dots_ocr.py [hunk 4]
```diff
@@ -0,0 +1,829 @@
...
+        self.embed_dim = dim
+        self.num_heads = num_heads
+        self.head_dim = dim // num_heads
+        self.tp_size = parallel_state.get_tensor_model_parallel_world_size()
+        self.tp_rank = parallel_state.get_tensor_model_parallel_rank()
+        self.num_heads_per_partition = dist_utils.divide(
+            num_heads, self.tp_size)
+
+        # qkv/proj follow Qwen2-VL style; bias controlled by arg
+        self.qkv = ColumnParallelLinear(input_size=dim,
```
Comment (jeejeelee) [inline vllm/model_executor/models/dots_ocr.py]: Maybe  we can use `QKVParallelLinear`.

**Comment by jeejeelee** [comment]
> > One potential performance issue: When I pass in 30 images, I see the below message 30 times before images are actually processed. This seems to add a big latency for this model.
> > ```
> > Fetching 1 files: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 7869.24it/s]
> > ```
> 
> Hmm - FYI @DarkLight1337, I'm wondering if this has something to do with a out-of-tree processor (i.e one with `trust_remote_code=True`) inheriting from a HF one?

I tested it with `vision_language.py` and didn't see similar output, which is a bit strange.

**Comment by casper-hansen** [comment]
> I tested it with `vision_language.py` and didn't see similar output, which is a bit strange.

I used `vllm serve rednote-hilab/dots.ocr --trust-remote-code` and just sent messages like so:

```
    async def _single_image_to_text(base64_image: str):
        messages = [
            {
                "role": "user",
                "content": [
                    {
                        "type": "image_url",
                        "image_url": {"url":  base64_image},
                    },
                    {"type": "text", "text": f"<|img|><|imgpad|><|endofimg|>Extract the text content from this image."}
                ],
            }
        ]

        response = await client.chat.completions.create(
            messages=messages, 
            model="rednote-hilab/dots.ocr", 
            max_completion_tokens=32768,
            temperature=0.1,
        )

        return response.choices[0].message.content
```

File: tests/models/registry.py [hunk 5]
```diff
@@ -447,6 +447,7 @@ def check_available_online(
                                                 max_transformers_version="4.48",  # noqa: E501
                                                 transformers_version_reason="HF model is not compatible.",  # noqa: E501
                                                 hf_overrides={"architectures": ["DeepseekVLV2ForCausalLM"]}),  # noqa: E501
+    "DotsOCRForCausalLM": _HfExamplesInfo("rednote-hilab/dots.ocr"),
```
Comment (ywang96) [inline tests/models/registry.py]: You'll need to add `trust_remote_code=True` for this I believe @yinz-aizip

**Comment by casper-hansen** [comment]
@ywang96 @yinz-aizip is it possible to avoid --trust-remote-code? I believe this is the root cause of the high latency.

**Comment by ywang96** [comment]
> I believe this is the root cause of the high latency.

@casper-hansen Yea I think so too - but I don't think it should be removed but instead we should cache the object that fetches the remote file (instead of doing it over and over) - This should not happen and I need to debug why this is happening 😅

File: vllm/model_executor/models/dots_ocr.py [hunk 6]
```diff
@@ -0,0 +1,851 @@
...
+        bias = config.use_bias
+
+        # Referenced aimv2.py AIMv2SwiGLUFFN
+        self.fc13 = MergedColumnParallelLinear(
+            in_features,
+            [hidden_features] * 2,
+            bias=bias,
+            quant_config=quant_config,
+            prefix=f"{prefix}.fc13",
+            disable_tp=True)
```
Comment (jeejeelee) [inline vllm/model_executor/models/dots_ocr.py]: I think we should pass `use_data_parallel` rather than hardcode, see the implementation of [Idefics2VisionTransformer](https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/idefics2_vision_model.py)

File: vllm/model_executor/models/dots_ocr.py [hunk 7]
```diff
@@ -0,0 +1,851 @@
...
+        else:
+            print("no norm in patch merger")
+
+        self.mlp = nn.Sequential(
+            nn.Linear(self.hidden_size, self.hidden_size),
+            nn.GELU(),
+            nn.Linear(self.hidden_size, dim),
+        )
+
+        if init_merger_std is not None:
```
Comment (jeejeelee) [inline vllm/model_executor/models/dots_ocr.py]: Will these codes still be used

File: vllm/model_executor/models/dots_ocr.py [hunk 8]
```diff
@@ -0,0 +1,851 @@
...
+            self.ln_q = LayerNorm(context_dim, eps=1e-6)
+        elif self.pre_norm == "rmsnorm":
+            self.ln_q = RMSNorm(context_dim, eps=1e-6)
+        else:
+            print("no norm in patch merger")
+
+        self.mlp = nn.Sequential(
+            nn.Linear(self.hidden_size, self.hidden_size),
+            nn.GELU(),
+            nn.Linear(self.hidden_size, dim),
```
Comment (jeejeelee) [inline vllm/model_executor/models/dots_ocr.py]: We can use `ReplicatedLinear` or parallel linear here

**Comment by DarkLight1337** [comment]
The repeated loading issue should be fixed by #25341

File: vllm/model_executor/models/dots_ocr.py (hunk 7 above)
Comment (ywang96) [reply vllm/model_executor/models/dots_ocr.py]: Looks like it's just training code - will remove

File: vllm/model_executor/models/dots_ocr.py (hunk 8 above)
Comment (ywang96) [reply vllm/model_executor/models/dots_ocr.py]: Cleaned it up

**Comment by ywang96** [comment]
The correctness of this PR has been verified by our contact from rednote engineering team so I'm just going to turn on `ready` label for it.

**Review by jeejeelee — APPROVED** [review]
Overall LGTM, some improvements can be completed in subsequent PRs
